rdflib/plugins/sparql/sparql.py

```diff
@@ -255,20 +255,39 @@ def __init__(
         graph: Optional[Graph] = None,
         bindings: Optional[Union[Bindings, FrozenBindings, List[Any]]] = None,
         initBindings: Optional[Mapping[str, Identifier]] = None,
+        datasetClause=None,
     ):
         self.initBindings = initBindings
         self.bindings = Bindings(d=bindings or [])
         if initBindings:
             self.bindings.update(initBindings)
 
         self.graph: Optional[Graph]
-        self._dataset: Optional[ConjunctiveGraph]
+        self._dataset: Optional[Dataset]
         if isinstance(graph, ConjunctiveGraph):
```

Comment:
The `isinstance(graph, ConjunctiveGraph)` statement on line below makes `mypy` think `graph` is a specifically a `ConjunctiveGraph` so it doesn't like the assignment to `self._dataset` that is typed `Optional[Dataset]`
So for now, until we strip out all of ConjunctiveGraph, lets make it `Union[Dataset, ConjunctiveGraph]`

```suggestion
        self._dataset: Optional[Union[Dataset, ConjunctiveGraph]]
        if isinstance(graph, (Dataset, ConjunctiveGraph)):
```